karabo_bridge/serializer.py

```diff
@@ -0,0 +1,203 @@
+            }
+
+        If the metadata dict is not provided it will be extracted from
+        'data' or an empty dict if 'metadata' key is missing from a data
+        source.
+
+    protocol_version: ('1.0' | '2.2')
+        Which version of the bridge protocol to use. Defaults to the latest
+        version implemented.
+
+    dummy_timestamps: bool
+        Some tools (such as OnDA) expect the timestamp information to be in the
+        messages. We can't give accurate timestamps where these are not in the
+        file, so this option generates fake timestamps from the time the data
+        is fed in, if the real timestamp information is missing.
+
+    returns
+    -------
+    msg: list of bytes/memoryviews ojects
+        binary conversion of data/metadata readable by the karabo_bridge
+    """
+    if protocol_version not in {'1.0', '2.2'}:
+        raise ValueError(f'Unknown protocol version {protocol_version}')
+
+    if metadata is None:
+        metadata = {src: v.get('metadata', {}) for src, v in data.items()}
+
+    if protocol_version == '1.0':
+        return _serialize_old(data, metadata, dummy_timestamps)
+
+    pack = msgpack.Packer(use_bin_type=True).pack
+    msg = []
+    ts = timestamp()
+    for src, props in sorted(data.items()):
+        src_meta = metadata[src].copy()
+        if dummy_timestamps and 'timestamp' not in src_meta:
+            src_meta.update(ts)
+
+        main_data = {}
+        arrays = []
+        for key, value in props.items():
+            if isinstance(value, np.ndarray):
+                arrays.append((key, value))
+            elif isinstance(value, np.number):
+                # Convert numpy type to native Python type
+                main_data[key] = value.item()
+            else:
+                main_data[key] = value
+
+        msg.extend([
+            pack({
+                'source': src, 'content': 'msgpack',
+                'metadata': src_meta
+            }),
+            pack(main_data)
+        ])
+
+        for key, array in arrays:
+            if not array.flags['C_CONTIGUOUS']:
+                array = np.ascontiguousarray(array)
+            msg.extend([
+                pack({
+                    'source': src, 'content': 'array', 'path': key,
+                    'dtype': str(array.dtype), 'shape': array.shape
+                }),
+                array.data,
+            ])
+
+    return msg
+
+
+def deserialize(msg):
+    """Deserializer for the karabo bridge protocol
+
+    Parameters
+    ----------
+    msg: list of zmq.Frame or list of byte objects
+        Serialized data following the karabo_bridge protocol
+
+    Returns
+    -------
+    data : dict
+        The data for a train, keyed by source name.
+    meta : dict
+        The metadata for a train, keyed by source name.
+    """
+    unpack = partial(msgpack.loads, raw=False, max_bin_len=0x7fffffff)
+
+    if not isinstance(msg[0], zmq.Frame):
+        msg = [Frame(m) for m in msg]
```

Comment:
What is the use case of this?

Reply:
in karabo bridge, we expect to receive `zmq.Frame` objects, because we read zeromq socket with no copy. the `serialize` function generates a list of binary object, not zeromq frames. This is useful if you want to deserialize data that you haven't received via the karabo bridge (for example if you use other kind of sockets)

Reply:
I see. This is not only for our bridge. Thank you!